### PR TITLE
[libc][stdbit] implement stdc_trailing_ones (C23)

### PR DESCRIPTION
**Pull request by nickdesaulniers**

### CODE REVIEW

**Comment by nickdesaulniers** [comment]
Note to reviewers, this commit is stacked on top of https://github.com/llvm/llvm-project/pull/80344. Only the commit "
[libc][stdbit] implement stdc_trailing_ones (C23)" (and subsequent pushes) should be reviewed.

File: libc/spec/stdc.td [hunk 1]
```diff
@@ -784,7 +784,17 @@ def StdC : StandardSpec<"stdc"> {
...
+          FunctionSpec<"stdc_trailing_zeros_uc", RetValSpec<UnsignedIntType>, [ArgSpec<UnsignedCharType>]>,
+          FunctionSpec<"stdc_trailing_zeros_us", RetValSpec<UnsignedIntType>, [ArgSpec<UnsignedShortType>]>,
+          FunctionSpec<"stdc_trailing_zeros_ui", RetValSpec<UnsignedIntType>, [ArgSpec<UnsignedIntType>]>,
+          FunctionSpec<"stdc_trailing_zeros_ul", RetValSpec<UnsignedIntType>, [ArgSpec<UnsignedLongType>]>,
+          FunctionSpec<"stdc_trailing_zeros_ull", RetValSpec<UnsignedIntType>, [ArgSpec<UnsignedLongLongType>]>,
+          FunctionSpec<"stdc_trailing_ones_uc", RetValSpec<UnsignedIntType>, [ArgSpec<UnsignedCharType>]>,
+          FunctionSpec<"stdc_trailing_ones_us", RetValSpec<UnsignedIntType>, [ArgSpec<UnsignedShortType>]>,
+          FunctionSpec<"stdc_trailing_ones_ui", RetValSpec<UnsignedIntType>, [ArgSpec<UnsignedIntType>]>,
+          FunctionSpec<"stdc_trailing_ones_ul", RetValSpec<UnsignedIntType>, [ArgSpec<UnsignedLongType>]>,
+          FunctionSpec<"stdc_trailing_ones_ull", RetValSpec<UnsignedIntType>, [ArgSpec<UnsignedLongLongType>]>
```
Comment (nickdesaulniers) [inline libc/spec/stdc.td:797]: should update Macros, too!

File: libc/include/llvm-libc-macros/stdbit-macros.h [hunk 2]
```diff
@@ -55,6 +85,20 @@ inline unsigned stdc_leading_ones(unsigned long long x) {
       unsigned: stdc_leading_ones_ui,                                          \
       unsigned long: stdc_leading_ones_ul,                                     \
       unsigned long long: stdc_leading_ones_ull)(x)
+#define stdc_trailing_zeros(x)                                                 \
+  _Generic((x),                                                                \
```
Comment (ThePhD) [inline libc/include/llvm-libc-macros/stdbit-macros.h:89]: Hey! Thanks for doing this. Please make sure that you include `_BitInt(CHAR_BIT)`/`_BitInt(CHAR_BIT * 2)`/`_BitInt(CHAR_BIT * 4)`/`_BitInt(CHAR_BIT * 8)` versions of this, as that's part of the spec. (You can also just use `8/16/32/64`; I know that LLVM as a thing tends to not care about `CHAR_BIT != 8` platforms.)

File: libc/test/src/stdbit/CMakeLists.txt [hunk 3]
```diff
@@ -1,112 +1,23 @@
...
-
+list(APPEND prefixes
+  leading_zeros
+  leading_ones
+  trailing_zeros
+  trailing_ones
+)
+list(APPEND suffixes c s i l ll)
+foreach(prefix IN LISTS prefixes)
+  foreach(suffix IN LISTS suffixes)
```
Comment (jhuber6) [inline libc/test/src/stdbit/CMakeLists.txt:11]:
```suggestion
foreach(prefix leading_zeros leading_ones trailing_zeros trailing_ones)
  foreach(suffix c s i l ll)
```
This should work unless we're actually depending on this appending to some list defined somewhere else.

File: libc/include/llvm-libc-macros/stdbit-macros.h (hunk 2 above)
Comment (nickdesaulniers) [reply libc/include/llvm-libc-macros/stdbit-macros.h:89]: > Hey! Thanks for doing this.

Heh, thanks for standardizing these.  I was going to show you/ask for review at some point, I appreciate that you found this, took a look, and had feedback.  Do you have cycles at the moment to be cc'ed on more of these as I implement them? If not, no worries.

This is the first I've heard of `_BitInt`, is that something new-ish?

> 6.2.5 Types
> A bit-precise signed integer type is designated as _BitInt(N) where N is an integer constant expression
that specifies the number of bits that are used to represent the type, including the sign bit. Each
value of N designates a distinct type.42)

Yeah, woah gcc-14 and clang-14 both support this type...https://godbolt.org/z/or83fMcro 

Ok, I will add support for these. Will file an TODO and revisit; the initial implementations don't need to be perfect to land.

File: libc/test/src/stdbit/CMakeLists.txt (hunk 3 above)
Comment (nickdesaulniers) [reply libc/test/src/stdbit/CMakeLists.txt:11]: Consider making this point in https://github.com/llvm/llvm-project/pull/80344 (the child PR this PR is based off of; sorry for stacking commits but this was more obvious in phabricator than github PRs).

The list of "suffixes" is set, but "prefixes" will grow to:
- leading_zeros
- leading_ones
- trailing_zeros
- trailing_ones
- first_leading_zero
- first_leading_one
- first_trailing_zero
- first_trailing_one
- count_zeros
- count_ones
- has_single_bit
- bit_width
- bit_floor
- bit_ceil

as I finish implementing stdbit.h.  Personal preference, but I prefer to add new individual lines for each of these to minimize the impact on `git blame`, as opposed to making one line excessively longer through multiple commits (which I believe is what you're suggesting).  I don't feel strongly about it, but that's an insight into *why* I did it that way. WDYT?

Comment (nickdesaulniers) [reply libc/test/src/stdbit/CMakeLists.txt:11]: > This should work unless we're actually depending on this appending to some list defined somewhere else.

Ah, and @lntue [pointed out](https://github.com/llvm/llvm-project/pull/80344#discussion_r1476788675) in review of the child patch that append is not necessary and to just use `set`.

File: libc/include/llvm-libc-macros/stdbit-macros.h (hunk 2 above)
Comment (nickdesaulniers) [reply libc/include/llvm-libc-macros/stdbit-macros.h:89]: Filed https://github.com/llvm/llvm-project/issues/80509 and will link to it from here.

File: libc/spec/stdc.td (hunk 1 above)
Comment (nickdesaulniers) [reply libc/spec/stdc.td:805]: 83c1e8912212

File: libc/include/llvm-libc-macros/stdbit-macros.h (hunk 2 above)
Comment (nickdesaulniers) [reply libc/include/llvm-libc-macros/stdbit-macros.h]: 9a4eac509ec4

**Review by lntue — APPROVED** [review]

**Pull request merged**